File: 백준/Silver/1780. 종이의 개수/종이의 개수.py
def d_q(matrix, row, col, size):
    global m, z, o

    ok = True
    target = matrix[row][col] #비교할 기준 숫자
    
    for i in range(row, row + size):
        for j in range(col, col +size):
            #i,j로 행렬 값 꺼내와서 값 비교하고 다르면 ok를 false로 바꾸고 break
            if matrix[i][j] != target:
                ok = False
                break
        if ok == False:
            break

    #ok가 true인 상태로 끝나면 m z o 에 해당하는거 +1
    if ok:
        if target == -1:
            m += 1
        elif target == 0:
            z += 1
        else:
            o += 1

    #ok가 false인 상태로 끝나면 d_q 재호출하면서 9개로 쪼개기 => 행 열 3으로 나누기
    else:
        new_size = size // 3
        for i in range(3):
            for j in range(3):
                d_q(matrix, row + i * new_size, col + j * new_size, new_size)

#근데 이게 분할정복 문제인지 잘 모르겠아요

#예제 입력 받기
n = int(input())
matrix = []

#입력값으로 매트릭스 만들기
for i in range(n):
    line = input()
    matrix.append(list(map(int,line.split())))

#프린트할 -1, 0, 1 개수 변수 초기화
m = z = o = 0

d_q(matrix, 0, 0, n)


print(m) #-1
print(z) #0
print(o) #1

File: 백준/Silver/1780. 종이의 개수/종이의 개수.cc
#include <iostream>
#include <vector>

using namespace std;

int m = 0, z = 0, o = 0;

void d_q(const vector<vector<int>>& matrix, int row, int col, int size) {
    int target = matrix[row][col]; // 비교할 기준 숫자
    bool ok = true;

    for (int i = row; i < row + size; ++i) {
        for (int j = col; j < col + size; ++j) {
            // 값이 다르면 ok를 false로 바꾸고 break
            if (matrix[i][j] != target) {
                ok = false;
                break;
            }
        }
        if (!ok) break;
    }

    if (ok) {
        if (target == -1) {
            m++;
        } else if (target == 0) {
            z++;
        } else {
            o++;
        }
    } else {
        int new_size = size / 3;
        for (int i = 0; i < 3; ++i) {
            for (int j = 0; j < 3; ++j) {
                d_q(matrix, row + i * new_size, col + j * new_size, new_size);
            }
        }
    }
}

int main() {
    int n;
    cin >> n;

    vector<vector<int>> matrix(n, vector<int>(n));
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n; ++j) {
            cin >> matrix[i][j];
        }
    }

    d_q(matrix, 0, 0, n);

    cout << m << endl;
    cout << z << endl;
    cout << o << endl;

    return 0;
}
